src/envs/playroom.py

```diff
@@ -0,0 +1,631 @@
+        self._robot_type = Type("robot", ["pose_x", "pose_y", "rotation",
+                                          "fingers"])
+        self._door_type = Type("door", ["pose_x", "pose_y", "open"])
+        self._dial_type = Type("dial", ["pose_x", "pose_y", "level"])
+        # Predicates
+        # Still need to add new predicates
+        self._On = Predicate(
+            "On", [self._block_type, self._block_type], self._On_holds)
+        self._OnTable = Predicate(
+            "OnTable", [self._block_type], self._OnTable_holds)
+        self._GripperOpen = Predicate(
+            "GripperOpen", [self._robot_type], self._GripperOpen_holds)
+        self._Holding = Predicate(
+            "Holding", [self._block_type], self._Holding_holds)
+        self._Clear = Predicate(
+            "Clear", [self._block_type], self._Clear_holds)
+        # Options
+        self._Pick = ParameterizedOption(
+            # variables: [robot, object to pick]
+            # params: [delta x, delta y, delta z]
+            "Pick", types=[self._robot_type, self._block_type],
+            params_space=Box(-1, 1, (3,)),
+            _policy=self._Pick_policy,
+            _initiable=self._Pick_initiable,
+            _terminal=self._Pick_terminal)
+        self._Stack = ParameterizedOption(
+            # variables: [robot, object on which to stack currently-held-object]
+            # params: [delta x, delta y, delta z]
+            "Stack", types=[self._robot_type, self._block_type],
+            params_space=Box(-1, 1, (3,)),
+            _policy=self._Stack_policy,
+            _initiable=self._Stack_initiable,
+            _terminal=self._Stack_terminal)
+        self._PutOnTable = ParameterizedOption(
+            # variables: [robot]
+            # params: [x, y] (normalized coordinates on the table surface)
+            "PutOnTable", types=[self._robot_type],
+            params_space=Box(0, 1, (2,)),
+            _policy=self._PutOnTable_policy,
+            _initiable=self._PutOnTable_initiable,
+            _terminal=self._PutOnTable_terminal)
+        # Still need to add new options
+        # Objects
+        self._robot = Object("robby", self._robot_type)
+        self._door1 = Object("door1", self._door_type)
+        self._door2 = Object("door2", self._door_type)
+        self._door3 = Object("door3", self._door_type)
+        self._door4 = Object("door4", self._door_type)
+        self._door5 = Object("door5", self._door_type)
+        self._door6 = Object("door6", self._door_type)
+        self._doors = (self._door1, self._door2, self._door3,
+                       self._door4, self._door5, self._door6)
+        self._dial = Object("dial", self._dial_type)
+
+    def simulate(self, state: State, action: Action) -> State:
+        assert self.action_space.contains(action.arr)
+        x, y, z, _, fingers = action.arr
+        # Update robot position
+        if not self._is_valid_loc(action):
```

Comment:
recommend changing this so that the input is the location only. I was initially confused because `action` isn't just a "loc".